### PR TITLE
JENKINS-67635 consider agent label expressions when applying trimLabels

### PR DESCRIPTION
**Pull request by NivKeidan**
See [JENKINS-67635](https://issues.jenkins-ci.org/browse/JENKINS-67635).

### Proposed changelog entries

* Launch only one agent to satisfy cloud agent requests that use label expressions.

### Submitter checklist

- [X] (If applicable) Jira issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by jtnord** [comment]
Looking at the bug report the behaviour seems correct and I am failing to understand what exactly the issue is.

When load balancing with labels a clid will be asked in the queue do you support x. If it says yes it will be used no further clouds are consulted 

This is a long standing limitation of the cloud API 
https://issues.jenkins.io/browse/JENKINS-57578 and causes other issues as well.

**Comment by timja** [comment]
(Behaviour was changed last in https://github.com/jenkinsci/jenkins/pull/5882 which was in 2.319.2)

**Comment by tobiasherzke** [comment]
> Looking at the bug report the behaviour reported as a bug actually seems correct to me and I am failing to understand what exactly the issue is.

I can see how the description in [JENKINS-67635](https://issues.jenkins.io/browse/JENKINS-67635) can be misunderstood. The issue is: 

Jenkins should spin up just one cloud agent and use it to execute the job. 

Instead, Jenkins spins up the first cloud agent, then does **not** use it to execute the job although it matches the label expression, instantiates another cloud agent, does **not** use it either, etc.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java:2270]: I am missing something here.
`Jenkins.getAssigneLabeles` returns a `Set<LabelAtom>`   
`getNodes()` returns a `List<Node>` and `Node.getAssignedLabeles()` also returns a `Set<LabelAtom>`

So whilst the Set of `nodeLabels` is declared as `Set<Label>` all the contents should be `LabelAtoms`

A LabelAtom should by definition be a single atomic label  so there should be no need for `l.matches(Colleciton)` here as a LabelAtom should only ever match the same Label Atom..

also `includedLabels.contains()` should be using equality for the set and `equals` is defined in Label to use `matches()` which works on the name of the label which is a string which in the case of LabelAtom is the label.

so if `n.getAssignedLabels` is not returning atoms of LabelAtoms then the code calling this on Node removal `void trimLabels(Node... nodes)` .

This would seem to point at something being wrong somewhere else?

Comment (JoeSSS) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: doesn't it make sense to revert https://github.com/jenkinsci/jenkins/pull/5882 for now then? most likely the problem is originated there

Comment (jtnord) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: alas not - as that fixed a major regression in the previous change here.

Comment (NivKeidan) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: maybe this is because the equals method in Label uses getClass for comparison?

Comment (jtnord) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: OK so `Label.equals(Object other)` (and `LabelAtom.equals` which defaults to the parent as it is not over ridden)
check class equality

https://github.com/jenkinsci/jenkins/blob/6edca13517f484e595f6f7341198239bc01ce8d5/core/src/main/java/hudson/model/Label.java#L514-L521

This will fail if anyone subclasses `LabelAtom`

Docker Swarm does 
https://github.com/jenkinsci/docker-swarm-plugin/blob/d6e0dc1fd4717dc178cf2ab81c8e7a5820fbf9fd/src/main/java/org/jenkinsci/plugins/docker/swarm/DockerSwarmLabelAssignmentAction.java#L49-L62

as does a few others.

Now `LabelAtoms` are supposed to be the actual Atom, and there may well be other cases where the code has assumed that LabelAtom is never subclassed when it is not.

Comment (jtnord) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: unless overridden in subclasses (which it is not as far as I can see) this means a Label subtype will only be equal to any other Label subtype if it's name matches, irrespective of any other things.

so whilst this code is correct there appears to be other potential bugs lurking here (and I still fail to see how for the ec2 plugin that is not subclassing LabelAtom that this fixes the issue.

Comment (NivKeidan) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: I am a maintainer of anka-build plugin and we are seeing similar issues (as explained in the issue related). We are also not overriding Label or LabelAtom, so this is not a specific EC2 thing (nor does it seem to have anything to do with overwriting these classes)

> includedLabels.contains() should be using equality for the set and equals is defined in Label to use matches() ...

I just ran some tests locally, the includedLabels.contains() does NOT fire the overridden equals method in Label

Comment (jtnord) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: > I just ran some tests locally, the includedLabels.contains() does NOT fire the overridden equals method in Label

contains is on a `HashSet` so the contains should be using the `hascode()` of the `LabelAtom` which is defined in `Label` as `name.hascode()`

Could you provide the example of the label that matches but is not in contains so I can understand this better?

Comment (NivKeidan) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: Furthermore, there can be cases, like the one we are discussing, where the equals method in Label would return false and not fire the underlying match method if comparing LabelAtom and LabelExpression.
Now, I am not sure if this behavior is expected or not, but wouldn't using 'instanceof Label' more appropriate in this equals implementation?

Comment (NivKeidan) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: > Could you provide the example of the label that matches but is not in contains so I can understand this better, as something seems wrong and I am trying to get to the root cause.

I'm sorry, I don't understand what you asking of me exactly.
You can see the exact label names and pipeline code I am using in the issue that this PR is trying to solve

Comment (jtnord) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: > > Could you provide the example of the label that matches but is not in contains so I can understand this better, as something seems wrong and I am trying to get to the root cause.
> 
> I'm sorry, I don't understand what you asking of me exactly. You can see the exact label names and pipeline code I am using in the issue that this PR is trying to solve

I do not have an ec2 cloud to test with - I have another cloud implementations and they appear to be working as expected.

when you said:

> I just ran some tests locally, the includedLabels.contains() does NOT fire the overridden equals method in Label

I possibly incorrectly assumed you had stepped though this code in a debugger with the cloud that is not behaving as you expected.  I was just asking you to provide the details of the `LabelAtom` that did not match in the `contains` but did match in the `matches`, along with the contents of the `includedLabels.

I will try and make time to setup the ec2 plugin and debug this to work out what is going on.

Comment (NivKeidan) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: Ahh, ok, is this what you mean?

I have one job running asking for 2 agents from label "lbl1||lbl2" in parallel
Four agents are being spun up, all of them lbl1

includedLabels: [cloud-lbl1-MPXKN, lbl1]
nodeLabels: [cloud-lbl1-gY3Hz, built-in, cloud-lbl1-MPXKN, cloud-lbl1-si1oa, lbl1, cloud-lbl1-7JsJ3]
l: lbl1||lbl2
contains returns false while matches returns true

**Review by jglick — COMMENTED** [review]
Clearly this will need test coverage.

**Comment by Vlatombe** [comment]
The patch looks correct to me. Manually checked that the reset on the label expression is really needed.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: ok, so walking this through with @Vlatombe the queue is not considering a task buildable as the labelExpression has no clouds (or nodes).  
This should fix that, but there is also another issue here in that label expressions are never garbage collected so long as something that can satisfy them exists, even if they where only used by a job/build that has long gone.  This is an existing issue and I will file an ticket for that.

At the same time it would seem that this is not limited to clouds but should also be able to be reproduced with a job and static agents (if the job had label expression `foo||bar`, and was run when Jenkins had no Nodes, then a user created a static node with the label `foo` - it would not build immediatly that node was added but would wait for the full maintainance task to kick in (`includedLabels == null`).

This should make a unit test possible without having to write a mock cloud implementation.

**Comment by timja** [comment]
> The patch looks correct to me. Manually checked that the reset on the label expression is really needed.

Can you expand a bit on what you mean? This patch or the previous one? What reset?

**Comment by Vlatombe** [comment]
> Can you expand a bit on what you mean? This patch or the previous one? What reset?

The comment from @jtnord has a bit more details.
https://github.com/jenkinsci/jenkins/pull/6193/files#r788853559

**Review by Vlatombe — APPROVED** [review]

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [reply core/src/main/java/jenkins/model/Jenkins.java:2270]: https://issues.jenkins.io/browse/JENKINS-67642 FTR

**Review by timja — APPROVED** [review]
Anyone got time to add test coverage here?

Seems like this fix is quite important quite a few people have asked about this

**Comment by NivKeidan** [comment]
I tried taking a look at the test code, and seemed pretty far off for me to implement a test for this.
Maybe if someone could give some pointers as to how to approach this and what test should be done, I could give it a try sometime this week

**Comment by NivKeidan** [comment]
hey guys, any update on this? Is this planned to be integrated into 2.319.3?

**Comment by timja** [comment]
> hey guys, any update on this? Is this planned to be integrated into 2.319.3?

It's not currently as it hasn't had enough testing, unless anyone objects we'll ship it in the next weekly version and backport it into the new LTS line due on 9th March

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by jglick** [comment]
> some pointers as to how to approach this and what test should be done

@Vlatombe / @jtnord anyone have some hints? I am not very familiar offhand with `Cloud` tests.

**Comment by jtnord** [comment]
DummyCLoudImpl exists and should be possible to use - but it would need to be [changed ](https://github.com/jenkinsci/jenkins-test-harness/blob/master/src/main/java/hudson/slaves/DummyCloudImpl.java#L103)to use label.matches other than that is should be a simple case of
create a cloud with a label X
create another cloud with label Y
create a job with restriction  X | Y
run a build and wait for success
assert there has been only one agent [provisioned](https://github.com/jenkinsci/jenkins-test-harness/blob/master/src/main/java/hudson/slaves/DummyCloudImpl.java#L61).

**Comment by timja** [comment]
> hey guys, any update on this? Is this planned to be integrated into 2.319.3?

were you able to have a go at the test @NivKeidan ?

**Comment by sithmein** [comment]
Can the fix be released before March for the LTS line? It's a regression that seriously affects everyone using dynamically allocated cloud agents.

**Comment by jtnord** [comment]
> Can the fix be released before March for the LTS line? It's a regression that seriously affects everyone using dynamically allocated cloud agents.

Fixes are eligible for [backporting](https://www.jenkins.io/download/lts/#backporting-process) if it would be in time for the `.1` I doubt (it has missed the cutoff to be in the weekly version that the LTS is cut from

**Comment by NivKeidan** [comment]
new test added

Fails without the fix suggested here and passes with it

**Review by tobiasherzke — COMMENTED** [review]
Thank you very much for the test. Question: Your test only succeeds if
`(1 == cloud1.numProvisioned && 0 == cloud2.numProvisioned)`
but
`(0 == cloud1.numProvisioned && 1 == cloud2.numProvisioned)`
would also be valid behaviour of Jenkins, right? And only some implementation details not relevant with regard to JENKINS-67635 may currently always produce the first result and not the second?
I may be wrong but I think

    assertTrue((1 == cloud1.numProvisioned && 0 == cloud2.numProvisioned) ||
               (0 == cloud1.numProvisioned && 1 == cloud2.numProvisioned));

would be a more robust test for this change.

**Comment by jtnord** [comment]
> * assert there has been only one agent [provisioned](https://github.com/jenkinsci/jenkins-test-harness/blob/master/src/main/java/hudson/slaves/DummyCloudImpl.java#L61) (one cloud should have one provisioned, the other should have zero)

Clouds are ordered, so if they can both provision only the first should provision (assuming provisioning needs to happen) so in the context of this test it should be ok.

**Pull request merged**
